prow/cmd/deck/job_history.go

```diff
@@ -0,0 +1,274 @@
+	dir, err := resolveSymLink(bkt, symLink)
+	if err != nil {
+		return &storage.ObjectHandle{}
+	}
+	return bkt.Object(path.Join(dir, fname))
+}
+
+func fileExists(bkt *storage.BucketHandle, root, id, fname string) bool {
+	obj := getObject(bkt, root, id, fname)
+	_, err := obj.Attrs(context.Background())
+	return err == nil
+}
+
+func readStarted(bkt *storage.BucketHandle, root, id string) (jobs.Started, error) {
+	s := jobs.Started{}
+	sobj := getObject(bkt, root, id, "started.json")
+	sdata, err := readObject(sobj)
+	if err != nil {
+		return s, fmt.Errorf("failed to read started.json for build %s: %v", id, err)
+	}
+	err = json.Unmarshal(sdata, &s)
+	if err != nil {
+		return s, fmt.Errorf("failed to parse started.json for build %s: %v", id, err)
+	}
+	return s, nil
+}
+
+func readFinished(bkt *storage.BucketHandle, root, id string) (jobs.Finished, error) {
+	f := jobs.Finished{}
+	fobj := getObject(bkt, root, id, "finished.json")
+	fdata, err := readObject(fobj)
+	if err != nil {
+		return f, fmt.Errorf("failed to read finished.json for build %s: %v", id, err)
+	}
+	err = json.Unmarshal(fdata, &f)
+	if err != nil {
+		return f, fmt.Errorf("failed to parse finished.json for build %s: %v", id, err)
+	}
+	return f, nil
+}
+
+// Gets job history from the GCS bucket specified in config.
+func getJobHistory(url *url.URL, config *config.Config, gcsClient *storage.Client) (JobHistoryTemplate, error) {
+	start := time.Now()
+
+	jobName := strings.TrimPrefix(url.Path, "/job-history/")
+	tmpl := JobHistoryTemplate{
+		Name:   jobName,
+		Builds: make([]BuildData, resultsPerPage),
+	}
+
+	var latest int
+	bucketName := config.ProwConfig.Plank.DefaultDecorationConfig.GCSConfiguration.Bucket
+	bkt := gcsClient.Bucket(bucketName)
+	var root string
+	found := false
+	for _, r := range []string{logsPrefix, symLinkPrefix} {
+		root = path.Join(r, jobName)
+		n, err := readLatestBuild(bkt, root)
+		if err == nil {
+			latest = n
+			found = true
+			break
+		}
+	}
+	if !found {
+		return tmpl, fmt.Errorf("failed to locate build data")
+	}
+	var top, bottom int // build ids of the top (inclusive) and bottom (exclusive) results
+	if idVals := url.Query()[idParam]; len(idVals) >= 1 {
+		var err error
+		if top, err = strconv.Atoi(idVals[0]); err != nil {
+			return tmpl, fmt.Errorf("invalid value for %s: %v", idParam, err)
+		}
+	} else {
+		top = latest
+	}
+	if top != latest {
+		newer := top + resultsPerPage
+		if newer > latest {
+			newer = latest
+		}
+		u := *url
+		q := u.Query()
+		q.Set(idParam, strconv.Itoa(newer))
+		u.RawQuery = q.Encode()
+		tmpl.NewerLink = u.String()
+		q.Set(idParam, strconv.Itoa(latest))
+		u.RawQuery = q.Encode()
+		tmpl.LatestLink = u.String()
+	}
+	bottom = top - resultsPerPage
+	// concurrently check if there are no older results to display
+	showOlder := false
+	fch := make(chan bool)
+	for i := bottom; i > bottom-resultsPerPage; i-- {
+		go func(i int) {
+			fch <- fileExists(bkt, root, strconv.Itoa(i), "started.json")
```

Comment:
This makes the assumption that build IDs are consecutive which is not the case unless `tot` is being used (which we want to avoid). BuildIDs look and act like timestamps if `tot` is not used.

Reply:
I fixed this in b6ed8475eb40a67bdcec941ec8a9acb68de57072, but it results in a huge performance hit for GCS directories with lots of objects.